### ISSUE TITLE
Chunk, window, attention_size

### ISSUE THREAD

**Issue by gzerveas** [maxjcohen/transformer#12]
Hello, nice work! Since this information is not currently included in the documentation, I was wondering what is exactly the role of chunks, how are they different from windows, and what does it mean when they are used together with an attention_size? Could you give some examples when one would want to use chunks vs windows vs attention_size?

**Comment by maxjcohen** [comment]
Hi, these are indeed important questions, that should be made clear in the documentation. I'll try to shed some light here.

Chunk and windows are two different implementation trying to solve the same limitation: the quadratic complexity in the temporal dimension of the Transformer. In other words, the Transformer performs slowly on long sequences. The idea behind both chunks and window MHA is to cut the input sequence into smaller, same sized sequences, and treat them in parallel.

- Chunk MHA simply splits the input sequence, causing potential lack of information at the beginning and ending of each chunk. This is illustrated in [this example](https://timeseriestransformer.readthedocs.io/en/latest/notebooks/trainings/training_2019_12_23__173446.html#Evaluate-on-the-test-dataset), where you can spot a higher loss at each end of the week (168 hours).
- Window MHA builds up on Chunk MHA by creating sub sequences using a moving window. In practice, this means that the model can look before and after the sub sequence to make a prediction, removing the spike in the loss at each chunk separation.

Attention size aims at helping the model focus on local patterns (i.e. the last few hours instead of the last few days), and does this by shutting of every input farther from `attention_size` at each time step (`attention_size` is usually 12 hours). This is done using the score matrix.

**TL;DR** Chunk and Window MHA are implementation tricks aimed at improving the Transformer computation speed on long sequences; Attention size aims at focusing attention on local patterns in order to improve performances.

**Comment by gzerveas** [comment]
Thank you very much for these answers! Just to clarify a couple of details:

- So using chunks seems to be equivalent to manually segmenting your input sequences into non-overlapping windows. Correct? I understand the issue at the beginning of chunks, but why would there be an issue close to the end of each chunk? For forecasting, isn't there a "look-ahead mask" which anyway prevents the model from looking into the future?

- Can windows be seen simply as "enlarged" chunks, expanded into the past and future? Wouldn't the "look-ahead mask" again prevent them from attending to the future, close to the end of the chunk? Or does the benefit come from the fact that during training the model is asked to do forecasting also for time points following each chunk?

- What is the intended advantage of this approach compared to a rolling, overlapping window? Is it just computational, or is it intended to make better use of a well-known periodicity of some time series (e.g. daily, monthly), and examine such periods in isolation of one another?

- Finally, I was also curious about what prompted you to implement transformers from scratch, as opposed to e.g. subclassing pytorch's transformers. Were there some limitations that couldn't be circumvented?

**Comment by maxjcohen** [comment]
- Yes precisely. Although the first attention block of the Decoder (the Self Attention) is prevented to attend to future information with a "look-ahead mask", the second attention block (Encoder Decoder Attention) does not have such a mask applied. Additionally, because we work on offline data (i.e. currently no forecasting), we considered removing the "look-ahead mask" altogether, and wanted to be prepared for this situation. In other words, since you need the entire week of data before processing the first hour, might as well look in the future, there might be interesting information.
- Yes you can look at it that way, and currently the extended part to the future probably has very little impact on the results. I have yet to test the impact on training.
- The Window MHA is a rolling overlapping window, just tweaked so that each segments is a week long. This is in fact because of the behavior of the dataset, which is sampled from a Building Energy Model (TRNSYS) which works by the week.
- There are two main reasons: firstly we wanted to be able to alter the Transformer as we saw fit, at any point during the process, secondly the pytorch's Transformer did not exist at the time of the creation of this repo :p

**Comment by gzerveas** [comment]
> 
> 
>     * Yes precisely. Although the first attention block of the Decoder (the Self Attention) is prevented to attend to future information with a "look-ahead mask", the second attention block (Encoder Decoder Attention) does not have such a mask applied. Additionally, because we work on offline data (i.e. currently no forecasting), we considered removing the "look-ahead mask" altogether, and wanted to be prepared for this situation. In other words, since you need the entire week of data before processing the first hour, might as well look in the future, there might be interesting information.
> 
>     * Yes you can look at it that way, and currently the extended part to the future probably has very little impact on the results. I have yet to test the impact on training.
> 
>     * The Window MHA is a rolling overlapping window, just tweaked so that each segments is a week long. This is in fact because of the behavior of the dataset, which is sampled from a Building Energy Model (TRNSYS) which works by the week.
> 
>     * There are two main reasons: firstly we wanted to be able to alter the Transformer as we saw fit, at any point during the process, secondly the pytorch's Transformer did not exist at the time of the creation of this repo :p

Thank you very much for your response! Good luck with the development effort! :)
